crates/uv/src/commands/project/ephemeral.rs

```diff
@@ -0,0 +1,125 @@
+        connectivity: Connectivity,
+        concurrency: Concurrency,
+        native_tls: bool,
+        cache: &Cache,
+        printer: Printer,
+    ) -> anyhow::Result<Self> {
+        let spec = RequirementsSpecification::from_requirements(requirements);
+
+        // Resolve the requirements with the interpreter.
+        let resolution = resolve_environment(
+            &interpreter,
+            spec,
+            settings.as_ref().into(),
+            state,
+            preview,
+            connectivity,
+            concurrency,
+            native_tls,
+            cache,
+            printer,
+        )
+        .await?;
+
+        // Hash the resolution by hashing the generated lockfile.
+        // TODO(charlie): If the resolution contains any mutable metadata (like a path or URL
+        // dependency), skip this step.
+        let lock = Lock::from_resolution_graph(&resolution)?;
+        let toml = lock.to_toml()?;
+        let resolution_hash = digest(&toml);
+
+        // Hash the interpreter by hashing the sysconfig data.
+        // TODO(charlie): Come up with a robust hash for the interpreter.
+        let interpreter_hash = digest(&interpreter.sys_executable());
+
+        // Search in the content-addressed cache.
+        let cache_entry = cache.entry(CacheBucket::Environments, interpreter_hash, resolution_hash);
+
+        // Lock the interpreter, to avoid concurrent modification across processes.
+        fs_err::tokio::create_dir_all(cache_entry.dir()).await?;
+        let _lock = LockedFile::acquire(
+            cache_entry.dir().join(".lock"),
+            cache_entry.dir().user_display(),
+        )?;
+
+        // If the receipt exists, return the environment.
+        let ok = cache_entry.path().join(".ok");
+        if ok.is_file() {
+            return Ok(Self(PythonEnvironment::from_root(
+                cache_entry.path(),
+                cache,
+            )?));
+        }
+
+        debug!(
+            "Creating ephemeral environment at: `{}`",
+            cache_entry.path().display()
+        );
+
+        let venv = uv_virtualenv::create_venv(
+            cache_entry.path(),
+            interpreter,
+            uv_virtualenv::Prompt::None,
+            false,
+            false,
+        )?;
+
+        // Install the ephemeral requirements.
+        // TODO(charlie): Rather than passing all the arguments to `sync_environment`, return a
+        // struct that lets us "continue" from `resolve_environment`.
+        let venv = sync_environment(
+            venv,
+            &resolution.into(),
+            settings.as_ref().into(),
+            state,
+            preview,
+            connectivity,
+            concurrency,
+            native_tls,
+            cache,
+            printer,
+        )
+        .await?;
+
+        // Create the receipt, to indicate to future readers that the environment is complete.
+        fs_err::tokio::File::create(ok).await?;
```

Comment:
So the receipt in this case is just an empty file? Kind of fun.

Reply:
Haha. We do the same thing (with the same name) in the Git checkouts so I just borrowed that pattern.

Reply:
Nice that seems like it might be a helpful pattern for more scoped locks in #4720